python/ray/tune/execution/tune_controller.py

```diff
@@ -460,7 +461,7 @@ def _maybe_add_actors(candidates: List[Trial]):
                 # If the trial is part of the list, but not of the set,
                 # we just ignore it. Removing it from the list on status
                 # change is too expensive.
-                if trial not in (self._pending_trials | self._paused_trials):
+                if trial not in self._pending_trials:
```

Comment:
We can theoretically have thousands of trials. The rationale here is that we add trials to `self._pending_trials`, which is a set and O(1), and to `self._pending_trials_list`, where insertion is O(1) and deletion is O(n). Instead of deleting from the list, we just iterate over the list and discard any objects that are not in the set.
The reason that we don't iterate over the set is because we want to keep the order the scheduler gave us intact.

Reply:
Oh, I mean this line, we're already incurring the O(n) deletion by popping from the front. Maybe we can change this to use a deque instead to have O(1) popleft.

https://github.com/ray-project/ray/blob/92f4fe725442a74705fe0c39d4cdf3f2d23d77c9/python/ray/tune/execution/tune_controller.py#L459